File: libraries/java/logs/src/main/java/com/cloudops/otel/logs/CloudOpsLogger.java
// This file contains cloud ops logger logic for OTel logs.
package com.cloudops.otel.logs;

import io.opentelemetry.api.OpenTelemetry;
import io.opentelemetry.api.baggage.Baggage;
import io.opentelemetry.api.common.Attributes;
import io.opentelemetry.api.common.AttributesBuilder;
import io.opentelemetry.api.logs.Logger;
import io.opentelemetry.api.logs.Severity;
import io.opentelemetry.api.trace.Span;
import io.opentelemetry.api.trace.SpanContext;
import io.opentelemetry.context.Context;
import io.opentelemetry.exporter.otlp.http.logs.OtlpHttpLogRecordExporter;
import io.opentelemetry.sdk.OpenTelemetrySdk;
import io.opentelemetry.sdk.common.CompletableResultCode;
import io.opentelemetry.sdk.logs.SdkLoggerProvider;
import io.opentelemetry.sdk.logs.export.BatchLogRecordProcessor;
import io.opentelemetry.sdk.logs.LogRecordProcessor;
import io.opentelemetry.sdk.resources.Resource;
import java.io.PrintStream;
import java.util.Arrays;
import java.util.EnumSet;
import java.util.Locale;
import java.util.Map;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.TimeUnit;

public final class CloudOpsLogger {
  private static final String[] DEFAULT_EXPORTERS = {"console"};
  private static volatile CloudOpsLogger instance;

  private final Set<LogLevel> enabledLevels;
  private final Map<String, String> resourceAttributes;
  private final String[] exportersList;
  private final LogSampler sampler;
  private SdkLoggerProvider loggerProvider;
  private Logger otelLogger;
  private boolean useConsole;
  private boolean useOtel;
  private String previousTraceId;
  private String uniqueId;

  CloudOpsLogger() {
    this.resourceAttributes = RuntimeResourceAttributes.create();
    this.enabledLevels = parseLogLevels(System.getenv("OTEL_LOG_LEVEL"));
    this.exportersList = LogsConfiguration.parseStringArray(System.getenv("OTEL_BACKEND_EXPORTERS"), DEFAULT_EXPORTERS);
    this.sampler = new LogSampler(this);
    init();
  }

  // Initializes logger.
  public static CloudOpsLogger initialiseLogger() {
    return initializeLogger();
  }

  // Initializes logger.
  public static CloudOpsLogger initializeLogger() {
    CloudOpsLogger localInstance = instance;
    if (localInstance == null) {
      synchronized (CloudOpsLogger.class) {
        localInstance = instance;
        if (localInstance == null) {
          localInstance = new CloudOpsLogger();
          instance = localInstance;
        }
      }
    }

    return localInstance;
  }

  // Handles resource attributes.
  public Map<String, String> resourceAttributes() {
    return resourceAttributes;
  }

  // Handles info.
  public void info(Object message, Object... optionalParams) {
    log(LogLevel.INFO, message, optionalParams);
  }

  // Handles error.
  public void error(Object message, Object... optionalParams) {
    log(LogLevel.ERROR, message, optionalParams);
  }

  // Handles debug.
  public void debug(Object message, Object... optionalParams) {
    log(LogLevel.DEBUG, message, optionalParams);
  }

  // Handles warn.
  public void warn(Object message, Object... optionalParams) {
    log(LogLevel.WARN, message, optionalParams);
  }

  // Exports logs.
  public void exportLogs() {
    sampler.flushOneBatch();
    if (useOtel && loggerProvider != null) {
      CompletableResultCode resultCode = loggerProvider.forceFlush();
      resultCode.join(30, TimeUnit.SECONDS);
    }
  }

  // Logs the requested work.
  private void log(LogLevel level, Object message, Object... optionalParams) {
    String currentTraceId = currentTraceId();
    if (!currentTraceId.equals(previousTraceId)) {
      uniqueId = UUID.randomUUID().toString();
      previousTraceId = currentTraceId;
    }

    sampler.addLog(new LogEntry(uniqueId == null ? "unknown" : uniqueId, level, message, optionalParams));
  }

  void processLog(LogEntry logEntry) {
    if (!enabledLevels.contains(logEntry.level)) return;

    String renderedMessage = render(logEntry.message, logEntry.optionalParams);
    if (useConsole) {
      PrintStream stream = logEntry.level == LogLevel.ERROR ? System.err : System.out;
      stream.println(renderedMessage);
    }

    if (useOtel && otelLogger != null) {
      otelLogger.logRecordBuilder()
          .setContext(Context.current())
          .setSeverity(toSeverity(logEntry.level))
          .setSeverityText(logEntry.level.value())
          .setBody(renderedMessage)
          .setAllAttributes(logAttributes(logEntry.invocationId))
          .emit();
    }
  }

  // Initializes the requested work.
  private void init() {
    if (exportersList.length <= 1 && containsExporter("console")) {
      useConsole = true;
      return;
    }

    SsmParameters ssmParameters = LogsConfiguration.readSsmParameters();
    if (ssmParameters == null || ssmParameters.isEmpty()) {
      useConsole = true;
      return;
    }

    for (String exporter : exportersList) {
      switch (exporter.trim().toLowerCase()) {
        case "console" -> useConsole = true;
        case "otel" -> initialiseOtel(ssmParameters);
        default -> useConsole = true;
      }
    }
  }

  // Initializes OTel.
  private void initialiseOtel(SsmParameters ssmParameters) {
    BackendConfig backendConfig = ssmParameters.backend("otel");
    if (backendConfig == null || backendConfig.logs == null) {
      useConsole = true;
      return;
    }

    var exporterBuilder = OtlpHttpLogRecordExporter.builder();
    if (LogsConfiguration.hasValue(backendConfig.logs.url)) {
      exporterBuilder.setEndpoint(backendConfig.logs.url);
    }
    if (LogsConfiguration.hasValue(backendConfig.logs.apiKey)) {
      exporterBuilder.addHeader("authorization", "Bearer " + backendConfig.logs.apiKey);
    }

    LogRecordProcessor processor = BatchLogRecordProcessor.builder(exporterBuilder.build()).build();
    loggerProvider = SdkLoggerProvider.builder()
        .setResource(Resource.getDefault().merge(resource()))
        .addLogRecordProcessor(processor)
        .build();

    OpenTelemetry openTelemetry = OpenTelemetrySdk.builder()
        .setLoggerProvider(loggerProvider)
        .build();
    otelLogger = openTelemetry.getLogsBridge().get(resourceAttributes.getOrDefault("service.name", "unknown_service"));
    useOtel = true;
  }

  // Handles resource.
  private Resource resource() {
    io.opentelemetry.sdk.resources.ResourceBuilder builder = Resource.builder();
    resourceAttributes.forEach(builder::put);
    return builder.build();
  }

  // Logs attributes.
  private Attributes logAttributes(String invocationId) {
    AttributesBuilder builder = Attributes.builder().put("invocation.id", invocationId == null ? "unknown" : invocationId);

    Baggage.current().forEach((key, entry) -> builder.put("baggage." + key, entry.getValue()));

    SpanContext spanContext = Span.current().getSpanContext();
    if (spanContext.isValid()) {
      builder.put("otel-trace-id", spanContext.getTraceId());
      builder.put("otel-span-id", spanContext.getSpanId());
    }

    return builder.build();
  }

  // Converts severity.
  private static Severity toSeverity(LogLevel level) {
    return switch (level) {
      case ERROR -> Severity.ERROR;
      case WARN -> Severity.WARN;
      case DEBUG -> Severity.DEBUG;
      case INFO -> Severity.INFO;
    };
  }

  // Parses log levels.
  static Set<LogLevel> parseLogLevels(String raw) {
    if (raw == null || raw.isBlank()) return EnumSet.allOf(LogLevel.class);

    EnumSet<LogLevel> levels = EnumSet.noneOf(LogLevel.class);
    for (String item : raw.replace("[", "").replace("]", "").replace("\"", "").split(",")) {
      String normalized = item.trim().toUpperCase(Locale.ROOT);
      if (normalized.isEmpty()) continue;

      try {
        levels.add(LogLevel.valueOf(normalized));
      } catch (IllegalArgumentException ignored) {
        // Ignore unknown level names so one bad value does not disable logging.
      }
    }

    return levels.isEmpty() ? EnumSet.allOf(LogLevel.class) : levels;
  }

  // Renders the requested work.
  private static String render(Object message, Object[] optionalParams) {
    String rendered = message instanceof Throwable throwable
        ? throwable.toString()
        : String.valueOf(message);

    if (optionalParams != null && optionalParams.length > 0) {
      rendered += "\n" + Arrays.deepToString(optionalParams);
    }

    return rendered;
  }

  // Handles contains exporter.
  private boolean containsExporter(String expected) {
    for (String exporter : exportersList) {
      if (expected.equalsIgnoreCase(exporter.trim())) return true;
    }

    return false;
  }

  // Gets current trace ID.
  private static String currentTraceId() {
    String lambdaTraceId = System.getenv("_X_AMZN_TRACE_ID");
    if (LogsConfiguration.hasValue(lambdaTraceId)) return lambdaTraceId;

    SpanContext spanContext = Span.current().getSpanContext();
    return spanContext.isValid() ? spanContext.getTraceId() : "unknown";
  }
}

File: libraries/dotnet/logs/src/CloudOpsLogger.cs
// This file contains cloud ops logger logic for logs src.
using System.Text;
using System.Text.Json;
using Microsoft.Extensions.Logging;
using OpenTelemetry.Exporter;
using OpenTelemetry.Logs;
using OpenTelemetry.Resources;

namespace CloudOps.Otel.Logs;

public sealed class CloudOpsLogger
{
    private static readonly IReadOnlyList<string> DefaultExporters = ["console"];
    private static readonly Lazy<CloudOpsLogger> Instance = new(() => new CloudOpsLogger());
    private readonly HashSet<LogLevel> enabledLevels;
    private readonly IReadOnlyList<string> exportersList;
    private readonly LogSampler sampler;
    private ILoggerFactory? loggerFactory;
    private ILogger? otelLogger;
    private bool useConsole;
    private bool useOtel;
    private string? previousTraceId;
    private string? uniqueId;

    // Handles cloud ops logger.
    internal CloudOpsLogger()
    {
        ResourceAttributes = RuntimeResourceAttributes.Create();
        enabledLevels = ParseLogLevels(Environment.GetEnvironmentVariable("OTEL_LOG_LEVEL"));
        exportersList = LogsConfiguration.ParseStringArray(Environment.GetEnvironmentVariable("OTEL_BACKEND_EXPORTERS"), DefaultExporters);
        sampler = new LogSampler(this);
        Init();
    }

    public IReadOnlyDictionary<string, string> ResourceAttributes { get; }

    // Initializes logger.
    public static CloudOpsLogger InitialiseLogger()
    {
        return Instance.Value;
    }

    // Initializes logger.
    public static CloudOpsLogger InitializeLogger()
    {
        return Instance.Value;
    }

    // Handles info.
    public void Info(object? message, params object?[] optionalParams)
    {
        Log(LogLevel.Info, message, optionalParams);
    }

    // Handles error.
    public void Error(object? message, params object?[] optionalParams)
    {
        Log(LogLevel.Error, message, optionalParams);
    }

    // Handles debug.
    public void Debug(object? message, params object?[] optionalParams)
    {
        Log(LogLevel.Debug, message, optionalParams);
    }

    // Handles warn.
    public void Warn(object? message, params object?[] optionalParams)
    {
        Log(LogLevel.Warn, message, optionalParams);
    }

    // Exports logs.
    public void ExportLogs()
    {
        sampler.FlushOneBatch();
        Console.Out.Flush();
        Console.Error.Flush();

        if (useOtel)
        {
            loggerFactory?.Dispose();
            InitOtelLogger();
        }
    }

    // Logs the requested work.
    private void Log(LogLevel level, object? message, object?[] optionalParams)
    {
        var currentTraceId = CurrentTraceId();
        if (currentTraceId != previousTraceId)
        {
            uniqueId = Guid.NewGuid().ToString();
            previousTraceId = currentTraceId;
        }

        sampler.AddLog(new LogEntry(uniqueId ?? "unknown", level, message, optionalParams));
    }

    // Processes log.
    internal void ProcessLog(LogEntry logEntry)
    {
        if (!enabledLevels.Contains(logEntry.Level)) return;

        var rendered = Render(logEntry.Message, logEntry.OptionalParams);
        if (useConsole)
        {
            if (logEntry.Level == LogLevel.Error)
            {
                Console.Error.WriteLine(rendered);
            }
            else
            {
                Console.WriteLine(rendered);
            }
        }

        if (useOtel && otelLogger is not null)
        {
            using var scope = otelLogger.BeginScope(LogAttributes(logEntry.InvocationId));
            otelLogger.Log(ToMicrosoftLevel(logEntry.Level), ToException(logEntry.Message), "{Message}", rendered);
        }
    }

    // Initializes the requested work.
    private void Init()
    {
        if (exportersList.Count <= 1 && exportersList.Contains("console", StringComparer.OrdinalIgnoreCase))
        {
            useConsole = true;
            return;
        }

        var exporterParameters = LogsConfiguration.ReadExporterParameters();
        if (exporterParameters.IsEmpty())
        {
            useConsole = true;
            return;
        }

        foreach (var exporter in exportersList.Select(item => item.Trim().ToLowerInvariant()))
        {
            switch (exporter)
            {
                case "console":
                    useConsole = true;
                    break;
                case "otel":
                    InitOtelLogger(exporterParameters);
                    break;
                default:
                    useConsole = true;
                    break;
            }
        }
    }

    // Initializes OTel logger.
    private void InitOtelLogger(ExporterParameters? exporterParameters = null)
    {
        exporterParameters ??= LogsConfiguration.ReadExporterParameters();
        var config = exporterParameters.Backend("otel")?.Logs;
        if (config is null)
        {
            useConsole = true;
            return;
        }

        loggerFactory = LoggerFactory.Create(builder =>
        {
            builder.SetMinimumLevel(Microsoft.Extensions.Logging.LogLevel.Trace);
            builder.AddOpenTelemetry(options =>
            {
                options.IncludeFormattedMessage = true;
                options.IncludeScopes = true;
                options.ParseStateValues = true;
                options.SetResourceBuilder(ResourceBuilder.CreateDefault().AddAttributes(
                    ResourceAttributes.Select(item => new KeyValuePair<string, object>(item.Key, item.Value))));
                options.AddOtlpExporter(exporterOptions =>
                {
                    exporterOptions.Protocol = OtlpExportProtocol.HttpProtobuf;
                    if (LogsConfiguration.HasValue(config.Url))
                    {
                        exporterOptions.Endpoint = new Uri(config.Url!);
                    }
                    var orgId = Environment.GetEnvironmentVariable("X_ORG_ID");
                    if (LogsConfiguration.HasValue(orgId))
                    {
                        exporterOptions.Headers = $"X-OrgId={orgId}";
                    }
                });
            });
        });

        otelLogger = loggerFactory.CreateLogger(ResourceAttributes.GetValueOrDefault("service.name", "unknown_service"));
        useOtel = true;
    }

    // Logs attributes.
    private static IReadOnlyDictionary<string, object> LogAttributes(string invocationId)
    {
        var attributes = new Dictionary<string, object>(StringComparer.Ordinal)
        {
            ["invocation.id"] = string.IsNullOrWhiteSpace(invocationId) ? "unknown" : invocationId
        };

        var activity = System.Diagnostics.Activity.Current;
        if (activity is not null)
        {
            if (activity.TraceId != default) attributes["otel-trace-id"] = activity.TraceId.ToString();
            if (activity.SpanId != default) attributes["otel-span-id"] = activity.SpanId.ToString();

            foreach (var baggage in activity.Baggage)
            {
                attributes[$"baggage.{baggage.Key}"] = baggage.Value ?? string.Empty;
            }
        }

        return attributes;
    }

    // Gets current trace ID.
    private static string CurrentTraceId()
    {
        var activity = System.Diagnostics.Activity.Current;
        return activity?.TraceId.ToString() ?? "unknown";
    }

    // Converts microsoft level.
    private static Microsoft.Extensions.Logging.LogLevel ToMicrosoftLevel(LogLevel level)
    {
        return level switch
        {
            LogLevel.Error => Microsoft.Extensions.Logging.LogLevel.Error,
            LogLevel.Warn => Microsoft.Extensions.Logging.LogLevel.Warning,
            LogLevel.Debug => Microsoft.Extensions.Logging.LogLevel.Debug,
            _ => Microsoft.Extensions.Logging.LogLevel.Information
        };
    }

    // Converts exception.
    private static Exception? ToException(object? message)
    {
        return message as Exception;
    }

    // Renders the requested work.
    private static string Render(object? message, object?[] optionalParams)
    {
        var builder = new StringBuilder()
            .Append(message is Exception exception ? exception.ToString() : Stringify(message));

        if (optionalParams.Length > 0)
        {
            builder.AppendLine().Append(string.Join(Environment.NewLine, optionalParams.Select(Stringify)));
        }

        return builder.ToString();
    }

    // Turns into text the requested work.
    private static string Stringify(object? value)
    {
        return value switch
        {
            null => "null",
            Exception exception => exception.ToString(),
            string text => text,
            _ when IsSimpleValue(value) => value.ToString() ?? string.Empty,
            _ => JsonSerializer.Serialize(value)
        };
    }

    // Checks whether simple value.
    private static bool IsSimpleValue(object value)
    {
        return value switch
        {
            bool or byte or sbyte or short or ushort or int or uint or long or ulong or float or double or decimal or char => true,
            _ => false
        };
    }

    // Parses log levels.
    internal static HashSet<LogLevel> ParseLogLevels(string? raw)
    {
        if (string.IsNullOrWhiteSpace(raw)) return Enum.GetValues<LogLevel>().ToHashSet();

        var levels = new HashSet<LogLevel>();
        foreach (var item in raw.Replace("[", string.Empty).Replace("]", string.Empty).Replace("\"", string.Empty).Split(','))
        {
            if (Enum.TryParse<LogLevel>(item.Trim(), ignoreCase: true, out var level))
            {
                levels.Add(level);
            }
        }

        return levels.Count > 0 ? levels : Enum.GetValues<LogLevel>().ToHashSet();
    }
}
